components/api_version/src/api_v2.rs

```diff
@@ -107,4 +108,65 @@ impl APIVersion for APIV2 {
         value.user_value.push(flags.bits());
         value.user_value
     }
+
+    fn decode_raw_key(encoded_key: &Key, with_ts: bool) -> Result<(Vec<u8>, Option<TimeStamp>)> {
+        debug_assert!(verify_encoded_key(encoded_key, with_ts));
+        let ts = decode_raw_key_timestamp(encoded_key, with_ts)?;
+        Ok((encoded_key.to_raw()?, ts))
+    }
+
+    fn decode_raw_key_owned(
+        encoded_key: Key,
+        with_ts: bool,
+    ) -> Result<(Vec<u8>, Option<TimeStamp>)> {
+        debug_assert!(verify_encoded_key(&encoded_key, with_ts));
+        let ts = decode_raw_key_timestamp(&encoded_key, with_ts)?;
+        Ok((encoded_key.into_raw()?, ts))
+    }
+
+    fn encode_raw_key(user_key: &[u8], ts: Option<TimeStamp>) -> Key {
+        debug_assert_eq!(Self::parse_key_mode(user_key), KeyMode::Raw);
+        let encoded_key = Key::from_raw(user_key);
+        if let Some(ts) = ts {
+            encoded_key.append_ts(ts)
+        } else {
+            encoded_key
+        }
+    }
+
+    fn encode_raw_key_owned(mut user_key: Vec<u8>, ts: Option<TimeStamp>) -> Key {
+        debug_assert_eq!(Self::parse_key_mode(&user_key), KeyMode::Raw);
+        let src_len = user_key.len();
+        let encoded_len = MemComparableByteCodec::encoded_len(src_len);
+        user_key.reserve(encoded_len - src_len + number::U64_SIZE * (ts.is_some() as usize));
```

Comment:
This is redundant as it will be called inside `resize`.

Reply:
We maybe need 8 more bytes for timestamp. Is it possible to extend the capacity again when we invoke `append_ts` later, if we do not reserve 8 more bytes here ?

Reply:
Got it.